### PR TITLE
Remove .index files after index sort

### PR DESCRIPTION
**Pull request by abudnik**
Changes:
1. .index file removed after index sort.
2. eblob_base_ctl struct refactoring: pack eblob_base_ctl data_{fd, offset, size} fields into data_ctl field.
3. removed extra sorted_index write & flush (before binlog apply).

### CODE REVIEW

File: library/datasort.c [hunk 1]
```diff
@@ -1475,7 +1469,7 @@ int eblob_generate_sorted_data(struct datasort_cfg *dcfg)
 	datasort_cleanup(dcfg);
 
 	/* Mark base as sorted */
-	dcfg->sorted_bctl->sorted = 1;
+	dcfg->sorted_bctl->data_ctl.sorted = 1;
```
Comment (shaitan) [inline library/datasort.c]: Should it also set `dcfg->sorted_bctl->index_ctl.sorted = 1;`?

Comment (shaitan) [inline library/datasort.c]: It's OK, I've missed that it is set in `datasort_swap_memory()`.

**Comment by shaitan** [comment]
:+1:  Looks good. @bioothod: have you reviewed it?

File: library/mobjects.c [hunk 2]
```diff
@@ -69,21 +69,23 @@ int eblob_base_setup_data(struct eblob_base_ctl *ctl, int force)
 	struct stat st;
 	int err;
 
-	err = fstat(ctl->index_fd, &st);
-	if (err) {
-		err = -errno;
-		goto err_out_exit;
+	if (ctl->index_ctl.fd >= 0) {
```
Comment (bioothod) [inline library/mobjects.c]: How is it possible not to have `index_ctl.fd >= 0` here?

File: library/mobjects.c [hunk 3]
```diff
@@ -300,53 +301,23 @@ static int eblob_base_ctl_open(struct eblob_backend *b, struct eblob_base_ctl *c
...
-
-			goto again;
-		}
-
-		ctl->index_fd = open(full, O_RDWR | O_CREAT | O_CLOEXEC, 0644);
-		if (ctl->index_fd < 0) {
-			err = -errno;
-			goto err_out_close_sort_fd;
-		}
+		unlink(full);
```
Comment (bioothod) [inline library/mobjects.c]: Does it remove old index, since everything lives in sorted index now?

Comment (abudnik) [inline library/mobjects.c]: Yes, it does.

File: library/mobjects.c (hunk 2 above)
Comment (abudnik) [inline library/mobjects.c]: eblob_base_ctl_open() calls eblob_base_setup_data() before initializing the index.

Comment (bioothod) [inline library/mobjects.c]: That code looks strange, `eblob_base_setup_data()` should always return error in this case and do not even try to open index files.

You hide this either error code or some unclear logic, which should be documented instead.

Comment (abudnik) [inline library/mobjects.c]: Fixed: https://github.com/abudnik/eblob/commit/ae8ba8e94557c445c33d664f605a4f8afa3d77c1

File: library/mobjects.c [hunk 4]
```diff
@@ -145,37 +143,32 @@ static int eblob_base_open_sorted(struct eblob_base_ctl *bctl, const char *dir_b
...
+	if (bctl->index_ctl.fd < 0) {
 		err = -errno;
 		goto err_out_free;
 	}
 
+	err = eblob_base_setup_data(bctl, 0);
+	if (err)
+		goto err_out_close;
+
+	if (bctl->index_ctl.size % sizeof(struct eblob_disk_control)) {
```
Comment (shaitan) [inline library/mobjects.c]: Can we move this check into `eblob_base_setup_data`?

Comment (abudnik) [inline library/mobjects.c]: https://github.com/abudnik/eblob/commit/a6de1f1dccf6e2ce443adda3680ed5961c8103fb

**Pull request merged**